core/client/src/main/java/alluxio/client/block/stream/PacketInStream.java

```diff
@@ -0,0 +1,278 @@
+  }
+
+  /**
+   * Creates a {@link PacketInStream} to read from a netty data server.
+   *
+   * @param context the file system context
+   * @param address the network address of the netty data server
+   * @param id the ID
+   * @param lockId the lock ID (set to -1 if not applicable)
+   * @param sessionId the session ID (set to -1 if not applicable)
+   * @param length the block or file length
+   * @param type the read request type (either block read or UFS file read)
+   * @return the {@link PacketInStream} created
+   */
+  public static PacketInStream createNettyPacketInStream(FileSystemContext context,
+      InetSocketAddress address, long id, long lockId, long sessionId, long length,
+      Protocol.RequestType type) {
+    PacketReader.Factory factory =
+        new NettyPacketReader.Factory(context, address, id, lockId, sessionId, type);
+    return new PacketInStream(factory, id, length);
+  }
+
+  /**
+   * Creates an instance of {@link PacketInStream}.
+   *
+   * @param packetReaderFactory the packet reader factory
+   * @param id the ID (either block ID or UFS file ID)
+   * @param length the length
+   */
+  public PacketInStream(PacketReader.Factory packetReaderFactory, long id, long length) {
+    mPacketReaderFactory = packetReaderFactory;
+    mId = id;
+    mLength = length;
+  }
+
+  @Override
+  public int read() throws IOException {
+    checkIfClosed();
+
+    readPacket();
+    if (mCurrentPacket == null) {
+      mEOF = true;
+    }
+    if (mEOF) {
+      close();
+      return -1;
+    }
+
+    mPos++;
+    return BufferUtils.byteToInt(mCurrentPacket.readByte());
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    checkIfClosed();
+    Preconditions.checkArgument(b != null, PreconditionMessage.ERR_READ_BUFFER_NULL);
+    Preconditions.checkArgument(off >= 0 && len >= 0 && len + off <= b.length,
+        PreconditionMessage.ERR_BUFFER_STATE.toString(), b.length, off, len);
+    if (len == 0) {
+      return 0;
+    }
+
+    readPacket();
+    if (mCurrentPacket == null) {
+      mEOF = true;
+    }
+    if (mEOF) {
+      close();
+      return -1;
+    }
+    int toRead = Math.min(len, mCurrentPacket.readableBytes());
+    mCurrentPacket.readBytes(b, off, toRead);
+    mPos += toRead;
+    return toRead;
+  }
+
+  @Override
+  public int positionedRead(long pos, byte[] b, int off, int len) throws IOException {
+    if (len == 0) {
+      return 0;
+    }
+    if (pos < 0 || pos >= mLength) {
+      return -1;
+    }
+
+    int lenCopy = len;
+    try (PacketReader reader = mPacketReaderFactory.create(pos, len)) {
+      // We try to read len bytes instead of returning after reading one packet because
+      // it is not free to create/close a PacketReader.
+      while (len > 0) {
+        ByteBuf buf = null;
+        try {
+          buf = reader.readPacket();
```

Comment:
consider to have ByteBufResource or PacketResource in the future to leverage try-with-resources.